pyquil/api/_qpu.py

```diff
@@ -71,34 +73,68 @@ def _extract_bitstrings(ro_sources: List[Optional[Tuple[int, int]]],
+        :param endpoint: Address to connect to the QPU server. If not provided, the
+            endpoint provided by engagement with dispatch is used. One or both must be
+            available and valid.
+        :param user: [deprecated] A string identifying who's running jobs.
+        :param priority: [deprecated] The priority with which to insert jobs into the QPU queue. Lower
                          integers correspond to higher priority.
+        :param config: PyQuilConfig object, which provides endpoint & engagement values.
         """
+        if config:
+            self.config = config
+        else:
+            self.config = PyquilConfig()
+
+        self.endpoint = endpoint
+        self._last_results: Dict[str, np.ndarray] = {}
+
+        super().__init__()
 
+    def build_client(self) -> Client:
+        endpoint = self.endpoint or self.config.qpu_url
         if endpoint is None:
-            raise RuntimeError("""It looks like you've tried to run a program against a QPU but do
- not currently have a reservation on one. To reserve time on Rigetti
- QPUs, use the command line interface, qcs, which comes pre-installed
- in your QMI. From within your QMI, type:
+            raise UserMessageError(
+                """It looks like you've tried to run a program against a QPU but do
+                not currently have a reservation on one. To reserve time on Rigetti
+                QPUs, use the command line interface, qcs, which comes pre-installed
+                in your QMI. From within your QMI, type:
 
-    qcs reserve --lattice <lattice-name>
+                    qcs reserve --lattice <lattice-name>
 
-For more information, please see the docs at
-https://www.rigetti.com/qcs/docs/reservations or reach out to Rigetti
-support at support@rigetti.com.""")
+                For more information, please see the docs at
+                https://www.rigetti.com/qcs/docs/reservations or reach out to Rigetti
+                support at support@rigetti.com.""")
 
-        self.client = Client(endpoint)
-        self.user = user
-        self._last_results: Dict[str, np.ndarray] = {}
-        self.priority = priority
+        logger.debug(f"QPU Client connecting to {endpoint}")
 
-        super().__init__()
+        return Client(endpoint, auth_config=self.client_auth_config)
+
+    @property
+    def client(self) -> Client:
+        if not (self.engagement and self.engagement.is_valid() and self._client):
+            self._client = self.build_client()
+        return self._client
+
+    @property
+    def client_auth_config(self) -> Optional[ClientAuthConfig]:
+        if self.engagement is not None:
+            return ClientAuthConfig(
+                client_public_key=self.engagement.client_public_key,
+                client_secret_key=self.engagement.client_secret_key,
+                server_public_key=self.engagement.server_public_key)
```

Comment:
same issue with properties -- if this is an important method than it should be a getter method and should be private if it is only used class-internal